### PR TITLE
Add checksum to the preload tarball download

### LINKED ISSUE #11061: VM starts but the download of `preloaded-images-k8s` was not completed
*(This is not a regular issue, but a support request)*

**Steps to reproduce the issue:** 
```
minikube start
minikube delete
minikube start
```
>😄  minikube v1.19.0 on Darwin 10.14.6
>    ▪ MINIKUBE_ACTIVE_DOCKERD=minikube
>✨  Using the virtualbox driver based on user configuration
>👍  Starting control plane node minikube in cluster minikube
>💾  Downloading Kubernetes v1.20.2 preload ...
>    \> preloaded-images-k8s-v10-v1...: 308.70 MiB / 491.71 MiB  62.78% 84.05 KiB
>🔥  Creating virtualbox VM (CPUs=3, Memory=5120MB, Disk=20000MB) ...
> ...

### Questions:

* Does Minikube has any way to *avoid re-downloading the `preloaded-images-k8s`, every time* that a VM is created?
* Where does Minikube stores the `preloaded-images-k8s`?\
So I can manually *copy it* before `delete` and *restore it* before `start`.

I am having connectivity problems and I would like to **reuse the images that were already downloaded on the first execution of `minikube start`**, when re-creating the Virtual Machine.

Is there **any** way to achieve this?

### PR DESCRIPTION
**Pull request by afbjorklund**
Check the checksum before putting the download into the cache,
leave the previous checksum file even if it is now redundant.

Closes #11061

### CODE REVIEW

File: pkg/minikube/download/preload.go [hunk 1]
```diff
@@ -141,6 +142,10 @@ func Preload(k8sVersion, containerRuntime string) error {
 	out.Step(style.FileDownload, "Downloading Kubernetes {{.version}} preload ...", out.V{"version": k8sVersion})
 	url := remoteTarballURL(k8sVersion, containerRuntime)
 
+	if checksum, err := getChecksum(k8sVersion, containerRuntime); err == nil {
+		url += "?checksum=" + checksum
+	}
```
Comment (medyagh) [inline pkg/minikube/download/preload.go:147]: how about adding a warning log when there is err is not nil and just log it for reference ?

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/minikube/download/preload.go (hunk 1 above)
Comment (afbjorklund) [reply pkg/minikube/download/preload.go:147]: we already break when failing to save the checksum, so might as well do the same here

File: pkg/minikube/download/preload.go [hunk 2]
```diff
@@ -141,6 +142,12 @@ func Preload(k8sVersion, containerRuntime string) error {
 	out.Step(style.FileDownload, "Downloading Kubernetes {{.version}} preload ...", out.V{"version": k8sVersion})
 	url := remoteTarballURL(k8sVersion, containerRuntime)
 
+	checksum, err := getChecksum(k8sVersion, containerRuntime)
+	if err != nil {
+		return errors.Wrap(err, "getting checksum")
```
Comment (medyagh) [inline pkg/minikube/download/preload.go:147]: should we return error here or should just at least try to download it regardless of being able to get checksum ?

Comment (afbjorklund) [reply pkg/minikube/download/preload.go:147]: It will fail a couple of lines down anyway, when it tries to save the same checksum to a file.

Comment (afbjorklund) [reply pkg/minikube/download/preload.go:147]: I don't know how unrealiable Google Cloud Storage, if it usually fails when trying to get metadata ?

Comment (afbjorklund) [reply pkg/minikube/download/preload.go]: Implemented the version without a checksum up front, doing it all manually instead (tempfile + verify)

**Comment by afbjorklund** [comment]
The code would look much better, if we just let go-getter do the tempfile and the validation.

File: pkg/minikube/download/preload.go [hunk 3]
```diff
@@ -141,6 +142,20 @@ func Preload(k8sVersion, containerRuntime string) error {
 	out.Step(style.FileDownload, "Downloading Kubernetes {{.version}} preload ...", out.V{"version": k8sVersion})
 	url := remoteTarballURL(k8sVersion, containerRuntime)
 
+	checksum, err := getChecksum(k8sVersion, containerRuntime)
+	var realPath string
+	if err != nil {
+		klog.Warningf("No checksum for preloaded tarball for k8s version %s", k8sVersion)
```
Comment (medyagh) [inline pkg/minikube/download/preload.go:148]: please log the error there, in addition to the k8sVersion

**Review by medyagh — CHANGES_REQUESTED** [review]

File: pkg/minikube/download/preload.go (hunk 3 above)
Comment (afbjorklund) [reply pkg/minikube/download/preload.go:148]: Done

**Review by medyagh — APPROVED** [review]

**Comment by medyagh** [comment]
Looks good to me, Thank you for this PR

**Comment by Pictor13** [comment]
Thank you for this fix! :)

**Pull request merged**